gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java

```diff
@@ -0,0 +1,505 @@
+    channelIsOpen = false;
+    closeContentChannel();
+  }
+
+  private void cacheFooter(ReadableByteChannel readableByteChannel) throws IOException {
+    footerContent = new byte[footerSize];
+    try (InputStream footerStream = Channels.newInputStream(readableByteChannel)) {
+      int totalBytesRead = 0;
+      int bytesRead;
+      do {
+        bytesRead = footerStream.read(footerContent, totalBytesRead, footerSize - totalBytesRead);
+        if (bytesRead >= 0) {
+          totalBytesRead += bytesRead;
+        }
+      } while (bytesRead >= 0 && totalBytesRead < footerSize);
+      checkState(
+          bytesRead >= 0,
+          "footerStream shouldn't be empty before reading the footer of size %s, totalBytesRead %s, read via last call %s, for '%s'",
+          footerSize,
+          totalBytesRead,
+          bytesRead,
+          resourceId);
+      checkState(
+          totalBytesRead == footerSize,
+          "totalBytesRead (%s) should equal footerSize (%s) for '%s'",
+          totalBytesRead,
+          footerSize,
+          resourceId);
+    } catch (IOException e) {
+      footerContent = null;
+      throw e;
+    }
+    logger.atFiner().log("Prefetched %s bytes footer for '%s'", footerContent.length, resourceId);
+  }
+
+  private ReadableByteChannel serveFooterContent() {
+    contentChannelPosition = currentPosition;
+    int offset = toIntExact(currentPosition - (objectSize - footerContent.length));
+    int length = footerContent.length - offset;
+    logger.atFiner().log(
+        "Opened channel (prefetched footer) from %d position for '%s'",
+        currentPosition, resourceId);
+    return Channels.newChannel(new ByteArrayInputStream(footerContent, offset, length));
+  }
+
+  private ReadableByteChannel openReadChannel(long bytesToRead) throws IOException {
+    checkArgument(bytesToRead > 0, "bytesToRead should be greater than 0, but was %s", bytesToRead);
+    checkState(
+        contentReadChannel == null && contentChannelEnd < 0,
+        "contentChannel and contentChannelEnd should be not initialized yet for '%s'",
+        resourceId);
+
+    if (footerContent != null && currentPosition >= objectSize - footerContent.length) {
+      return serveFooterContent();
+    }
+
+    contentChannelPosition = getRangeRequestStart();
+    long rangeLength = getRangeRequestLength(bytesToRead);
+    contentChannelEnd = rangeLength == -1 ? objectSize : contentChannelPosition + rangeLength;
+    checkState(
+        contentChannelEnd >= contentChannelPosition,
+        String.format(
+            "Start position should be <= endPosition startPosition:%d, endPosition: %d",
+            contentChannelPosition, contentChannelEnd));
+
+    ReadableByteChannel readableByteChannel =
+        getStorageReadChannel(contentChannelPosition, contentChannelEnd);
+
+    // Check if footer was fetched
+    if (contentChannelEnd == objectSize
+        && contentChannelEnd - contentChannelPosition == footerSize) {
```

Comment:
Please put "contentChannelEnd - contentChannelPosition" in brackets for better readability

Reply:
Done.